routing/cross_mwm_graph.cpp

```diff
@@ -0,0 +1,214 @@
+            minDistSegs[numMwmId].m_minDistTwinSeg = tc;
+          }
+        }
+      }
+    };
+
+    m_index.ForEachInRect(findBestTwins,
+                          MercatorBounds::RectByCenterXYAndSizeInMeters(p, kTransitionEqualityDistM),
+                          scales::GetUpperScale());
+
+    for (auto const & kv : minDistSegs)
+    {
+      if (kv.second.m_exactMatchFound)
+        continue;
+      if (kv.second.m_minDistM == kInvalidDistance)
+        continue;
+      twins.push_back(kv.second.m_minDistTwinSeg);
+    }
+  }
+
+  my::SortUnique(twins);
+
+  for (Segment const & t : twins)
+    CHECK_NOT_EQUAL(s.GetMwmId(), t.GetMwmId(), ());
+}
+
+void CrossMwmGraph::GetEdgeList(Segment const & s, bool isOutgoing, vector<SegmentEdge> & edges)
+{
+  CHECK(IsTransition(s, !isOutgoing), ("The segment is not a transition segment. IsTransition(",
+                                       s, "," , !isOutgoing, ") returns false."));
+  edges.clear();
+
+  // Osrm based cross-mwm information.
+
+  // Note. According to cross-mwm OSRM sections there're node id which could be ingoing and outgoing
+  // at the same time. For example in Berlin mwm on Nordlicher Berliner Ring (A10) near crossing
+  // with A11
+  // there's such node id. It's an extremely rare case. There're probably several such node id
+  // for the whole Europe. Such cases are not processed in WorldGraph::GetEdgeList() for the time
+  // being.
+  // To prevent filling |edges| with twins instead of leap edges and vice versa in
+  // WorldGraph::GetEdgeList()
+  // CrossMwmGraph::GetEdgeList() does not fill |edges| if |s| is a transition segment which
+  // corresponces node id described above.
+  if (IsTransition(s, isOutgoing))
+    return;
+
+  // Index graph based cross-mwm information.
+  if (CrossMwmSectionExists(s.GetMwmId()))
+    m_crossMwmIndexGraph.GetEdgeList(s, isOutgoing, edges);
+  else
+    m_crossMwmOsrmGraph.GetEdgeList(s, isOutgoing, edges);
+}
+
+void CrossMwmGraph::Clear()
+{
+  m_crossMwmOsrmGraph.Clear();
+  m_crossMwmIndexGraph.Clear();
+}
+
+TransitionPoints CrossMwmGraph::GetTransitionPoints(Segment const & s, bool isOutgoing)
+{
+  if (CrossMwmSectionExists(s.GetMwmId()))
+  {
+    CrossMwmConnector const & connector = m_crossMwmIndexGraph.GetCrossMwmConnectorWithTransitions(s.GetMwmId());
+    // In case of transition segments of index graph cross-mwm section the front point of segment
+    // is used as a point which corresponds to the segment.
+    return TransitionPoints({connector.GetPoint(s, true /* front */)});
+  }
+  return m_crossMwmOsrmGraph.GetTransitionPoints(s, isOutgoing);
+}
+
+bool CrossMwmGraph::CrossMwmSectionExists(NumMwmId numMwmId)
+{
+  if (m_crossMwmIndexGraph.HasCache(numMwmId))
+    return true;
+
+  MwmValue * value = m_index.GetMwmHandleByCountryFile(m_numMwmIds->GetFile(numMwmId)).GetValue<MwmValue>();
+  CHECK(value != nullptr, ("Country file:", m_numMwmIds->GetFile(numMwmId)));
```

Comment:
You still need to check that mwm is alive. It's possible. Try to take handle for non-existing mwm, for example :)